### ISSUE TITLE
Parameter override using environmental variables

### ISSUE THREAD

**Issue by danceb** [qwc-services/qwc-config-generator#83]
In different other services "config options in the config file can be overridden by equivalent uppercase environment variables."
I thought, this should be possible in all services, but this does not seems to be the case.
So it is possible to get this functionality in all the other services, too? Right in the moment I would use this within the qwc-config-generator, but I think to match this, it would be perfect to get this possibility within all the services.

Thanks in advance!

**Comment by manisandro** [comment]
Do you have an example where this does not apply? The env-override is handled generically for all config variables here [1].

[1] https://github.com/qwc-services/qwc-services-core/blob/master/qwc_services_core/runtime_config.py#L55

**Comment by danceb** [comment]
oh ok, so I guessed this right.

I have tried several config parameters.

from the tenantConfig:
```yaml
  "service": "config-generator",
  "config": {
    "default_qgis_server_url": "http://qwc-qgis-server/ows/",
    "qgis_print_layouts_dir": "/layouts",
    [...]
  },
```
from the docker-compose.yml:
```yaml
  qwc-config-service:
    image: sourcepole/qwc-config-generator:latest
    restart: unless-stopped
    environment:
      <<: *qwc-service-variables
      DEFAULT_QGIS_SERVER_URL: "http://example.com/ows/"
      QGIS_PRINT_LAYOUTS_DIR: "/anotherdir"
      [...]
```
The variables are set correct within the container:
```sh
> docker exec -it qwc-config-service-1 sh
> / # printenv
[...]
DEFAULT_QGIS_SERVER_URL=http://example.com/ows/
QGIS_PRINT_LAYOUTS_DIR=/anotherdir
```
Creating config files with:
```bash
docker exec -t qwc-config-service-1 python3 /srv/qwc_service/config_generator_cli.py /srv/qwc_service/config-in/default/tenantConfig.json all
```
From the logs:
```
QWC ConfigGenerator
[2024-12-10 22:08:23.781112] DEBUG: Using tenant 'default'
[2024-12-10 22:08:23.782411] INFO: Config destination: '/srv/qwc_service/config-out/default'
[...]
[2024-12-10 22:08:25.034186] INFO: <b>Searching for print layouts in /layouts</b>
[2024-12-10 22:08:25.045321] INFO: Found print template ...
[...]
[2024-12-10 22:08:25.149236] INFO: <b>Reading theme /ows/pg/qwc_config/alkis</b>
[2024-12-10 22:08:26.165518] INFO: Downloaded WMS GetProjectSettings from http://qwc-qgis-server/ows/pg/qwc_config/alkis
[...]
```
So instead of the parameters from the environment variables, it tooks the config parameters from the tenantConfig.json.

**Comment by manisandro** [comment]
Can you try adding some prints to the file [1] in the container to see what is going on?

[1] https://github.com/qwc-services/qwc-services-core/blob/master/qwc_services_core/runtime_config.py#L55

**Comment by danceb** [comment]
Maybe this is a special case, as the config parameters for the config generator are not part of the `services` block in the `tenantConfig.json`?
```json
{
  "$schema": "https://github.com/qwc-services/qwc-config-generator/raw/master/schemas/qwc-config-generator.json",
  "service": "config-generator",
  "config": {
    [...]
  },
  "themesConfig": "./themesConfig.json",
  "services": [
    {
      [...]
    }
  ]
}
```

**Comment by danceb** [comment]
So all services are identified by there "name" tag in [1] and I think all of the service generation steps are applied on these services. 
Besides this the generator config is handled seperatly and there is no option to get the environmental variable version instead [2].



The need in my case is to override the qgis server url, so
```python
        self.default_qgis_server_url = generator_config.get(
            'default_qgis_server_url',  os.environ.get(
               'DEFAULT_QGIS_SERVER_URL', 'http://localhost:8001/ows/'
               )).rstrip('/') + '/'
```
instead of 
```python
        self.default_qgis_server_url = generator_config.get(
            'default_qgis_server_url', 'http://localhost:8001/ows/'
        ).rstrip('/') + '/'
```
in [3] should do the thing, right?

But maybe there is a better solution to do the override in general for all of the generator config parameters.

[1] https://github.com/qwc-services/qwc-config-generator/blob/master/src/config_generator/config_generator.py#L182
[2] https://github.com/qwc-services/qwc-config-generator/blob/master/src/config_generator/config_generator.py#L197
[3] https://github.com/qwc-services/qwc-config-generator/blob/master/src/config_generator/config_generator.py#L199-201

**Comment by manisandro** [comment]
Looking somewhat in more detail at the code, at this stage `config` is a simple python dict, especially after merging the `tenantConfig.template.json`. Probably an environment variable should have precedence over any merged config, but this currently won't be the case. Possibly adding a 

    class RuntimeConfig:
        [...]
        def from_config(self, config):
            self.config = config
            return self
        [...]

method and then writing

    generator_config = RuntimeConfig.from_config(config)
    
could make it work again.

**Comment by danceb** [comment]
Ok. As I don't have a develop environment with `qwc-services-core`, could you please give it a try?

**Comment by manisandro** [comment]
You can try editing the files directly in the docker container and then restarting container or (still inside the container)

    kill $(pidof uwsgi)

to restart uswgi.

**Comment by danceb** [comment]
Thanks to give me the possibility to test such things by myself (learning by doing), but where could I find the qwc-services-core files within the container of the config generator?

**Comment by manisandro** [comment]
You can locate it i.e. by

    $ find / -name runtime_config.py
    /usr/lib/python3.11/site-packages/qwc_services_core/runtime_config.py

**Comment by danceb** [comment]
Yippieh, it worked!

Besides your suggestion
```python
class RuntimeConfig:
    [...]
    def from_config(self, config):
        self.config = config
        return self
    [...]
```
in `qwc_services_core.runtime_config` it needs of couse the related import statement in `config_generator.py`:
```python
from qwc_services_core.runtime_config import RuntimeConfig
[...]
generator_config = RuntimeConfig(None, logger).from_config(config)
```

Would be awesome, if could implement the changes soon! Thank you very much!

**Comment by manisandro** [comment]
Done in v2024.12.11 (note that you will also need to update the other images if you update this one [1])

[1] https://qwc-services.github.io/master/release_notes/QwcDockerUpgradeNotes/#qwc-services-v20241118-mass-update-update

**Comment by danceb** [comment]
Thanks!

**Comment by danceb** [comment]
Unfortunately there is something missing ...

With the new adaption the config generation throws an error, if the optional `qgis_projects_scan_base_dir` parameter missing (I don't get it, why exactly this one...)

```
Traceback (most recent call last):
  File "/srv/qwc_service/config_generator_cli.py", line 60, in <module>
    generator = ConfigGenerator(config, logger, os.path.dirname(args.config_file), False, False)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/qwc_service/config_generator/config_generator.py", line 309, in __init__
    'mapViewer': MapViewerConfig(
                 ^^^^^^^^^^^^^^^^
  File "/srv/qwc_service/config_generator/map_viewer_config.py", line 103, in __init__
    'qgis_projects_scan_base_dir').rstrip('/') + '/'
                                   ^^^^^^
AttributeError: 'NoneType' object has no attribute 'rstrip'
```

**Comment by manisandro** [comment]
Should be fixed in v2024.12.12
